drivers/can/stm32_ll_can.c

```diff
@@ -0,0 +1,725 @@
+	k_sem_init(&data->mb1.tx_int_sem, 0, 1);
+	k_sem_init(&data->mb2.tx_int_sem, 0, 1);
+	data->mb0.tx_callback = NULL;
+	data->mb1.tx_callback = NULL;
+	data->mb2.tx_callback = NULL;
+
+	data->filter_usage = (1ULL << CAN_MAX_NUMBER_OF_FILTES) - 1ULL;
+	memset(data->rx_response, 0, sizeof(void *) * CONFIG_CAN_MAX_FILTER);
+	data->response_type = 0;
+
+	clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	__ASSERT_NO_MSG(clock);
+
+	ret = clock_control_on(clock, (clock_control_subsys_t *) &cfg->pclken);
+	if (ret != 0) {
+		SYS_LOG_ERR("HAL_CAN_Init clock control on failed: %d\n", ret);
+		return -EIO;
+	}
+
+	ret = can_stm32_runtime_configure(dev, can_mode_normal, 0);
+	if (ret) {
+		return ret;
+	}
+
+	cfg->config_irq(can);
+	can->IER |= CAN_IT_TME;
+	SYS_LOG_INF("Init of %s done\n", dev->config->name);
+	return 0;
+}
+
+int can_stm32_runtime_configure(struct device *dev, can_mode_t mode,
+				u32_t bitrate)
+{
+	CAN_HandleTypeDef hcan;
+	const struct can_stm32_config *cfg = DEV_CFG(dev);
+	CAN_TypeDef *can = cfg->can;
+	struct device *clock;
+
+	clock = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	__ASSERT_NO_MSG(clock);
+	hcan.Instance = can;
+
+	u32_t clock_rate;
+	clock_control_get_rate(clock, (clock_control_subsys_t *) &cfg->pclken,
+			       &clock_rate);
+	if (!bitrate) {
+		bitrate = cfg->bus_speed;
+	}
+	u32_t prescaler = clock_rate / (BIT_SEG_LENGTH * cfg->bus_speed);
+	if (prescaler == 0 || prescaler > 1024) {
+		SYS_LOG_ERR("HAL_CAN_Init failed: prescaler > max (%d > 1024)\n",
+			    prescaler);
+		return -EINVAL;
+	}
+
+	const u32_t bs1 = (CONFIG_CAN_PHASE_SEG1_PROP_SEG - 1) << CAN_BTR_TS1_Pos;
+	const u32_t bs2 = (CONFIG_CAN_PHASE_SEG2 - 1) << CAN_BTR_TS2_Pos;
+	const u32_t swj = (CONFIG_CAN_SJW - 1) << CAN_BTR_SJW_Pos;
+	u32_t hal_mode  = mode == can_mode_normal   ? CAN_MODE_NORMAL   :
+			  mode == can_mode_loopback ? CAN_MODE_LOOPBACK :
+			  mode == can_mode_silent   ? CAN_MODE_SILENT   :
+			  CAN_MODE_SILENT_LOOPBACK;
+
+	hcan.Init.TTCM = DISABLE;
+	hcan.Init.ABOM = DISABLE;
+	hcan.Init.AWUM = DISABLE;
+	hcan.Init.NART = DISABLE;
+	hcan.Init.RFLM = DISABLE;
+	hcan.Init.TXFP = DISABLE;
+	hcan.Init.Mode = hal_mode;
+	hcan.Init.SJW  = swj;
+	hcan.Init.BS1  = bs1;
+	hcan.Init.BS2  = bs2;
+	hcan.Init.Prescaler = prescaler;
+
+	int hal_ret = HAL_CAN_Init(&hcan);
+	if (hal_ret != HAL_OK) {
+		SYS_LOG_ERR("HAL_CAN_Init failed: %d\n", hal_ret);
+		return -EIO;
+	}
+	SYS_LOG_DBG("Runtime configure of %s done\n", dev->config->name);
+	return 0;
+}
+
+int can_stm32_send(struct device *dev, struct can_msg *msg, s32_t timeout,
+		   can_tx_callback_t callback)
+{
+	const struct can_stm32_config *cfg = DEV_CFG(dev);
+	struct can_stm32_data *data = DEV_DATA(dev);
+	CAN_TypeDef *can = cfg->can;
+	u32_t transmit_status_register = can->TSR;
+	CAN_TxMailBox_TypeDef *mailbox = NULL;
+	struct k_mutex *tx_mutex = &data->tx_mutex;
+	struct can_mailbox *mb = NULL;
+
+	SYS_LOG_DBG("Sending %d bytes on %s\n"
```

Comment:
Normally syslog prints should not have \n in them as that will look funny if the syslog is processed remotely. I think you could condense the line and have everything in one line. Note that syslog will automatically place \n at the end of the printed line if printing to console so you do not have to put it there.